torchtitan/parallelisms/parallel_dims.py

```diff
@@ -18,36 +18,53 @@ class ParallelDims:
     pp: int
     world_size: int
     enable_loss_parallel: bool
-    dp_type: str
+    dp_type: str = "fsdp"
+    dp_replicate: int = 1  # Only used when dp_type is hsdp
 
     def __post_init__(self):
         self.dp_type = self.dp_type.lower()
         self._validate()
 
     def _validate(self):
-        dp, tp, pp = self.dp, self.tp, self.pp
+        dp, dp_replicate, tp, pp = self.dp, self.dp_replicate, self.tp, self.pp
         if dp == -1:
             self.dp = dp = self.world_size // (tp * pp)
         assert dp >= 1, dp
+        assert dp_replicate >= 1 and dp % dp_replicate == 0, (dp, dp_replicate)
         assert tp >= 1, tp
         assert pp >= 1, pp
         assert (
-            dp * tp * pp == self.world_size
+            dp * tp * pp == self.world_size,
         ), f"Invalid parallel dims: dp({dp}) * tp({tp}) * pp({pp}) != WORLD_SIZE({self.world_size})"
-        assert self.dp_type in ("fsdp", "ddp")
+        assert self.dp_type in ("fsdp", "ddp", "hsdp")
+        assert self.dp_type != "hsdp" or dp_replicate > 1, (self.dp_type, dp_replicate)
 
     def build_mesh(self, device_type):
         dims = []
         names = []
         for d, name in zip(
             [self.pp, self.dp, self.tp], ["pp", "dp", "tp"], strict=True
         ):
-            if d > 1:
+            if d <= 1:
+                continue
```

Comment:
hm, does this change existing behavior, e.g. we used to have pp=1 build a meshdim but now we do not?

this may be OK, just checking